### PR TITLE
Fix accidental non-collectible context unloading

### LINKED ISSUE #22902: The AssemblyLoadContext method calls the LoadFromStream method at runtime to throw an AssemblyLoadContext_Verify_NotUnloading exception.
Hello .Net Core Community,
[Surging](https://github.com/dotnetcore/surging)  Is a service engine framework based on dotnetcore. It has a problem for a long time. Can you help me? Because need to make the business module hot update, need to generate a dynamic proxy at runtime to achieve grayscale publishing, but use AssemblyLoadContext hot The update loader throws an exception `AssemblyLoadContext is unloading or was already unloaded.`, the code  follows:
```c#
        var stream = CompilationUtilitys.CompileClientProxy(trees,
                assemblys
                    .Select(a => MetadataReference.CreateFromFile(a.Location))
                    .Concat(new[]
                    {
                        MetadataReference.CreateFromFile(typeof(Task).GetTypeInfo().Assembly.Location)
                    }),
                _logger);

            using (stream)
            {
#if NET
                var assembly = Assembly.Load(stream.ToArray());
#else
                var assembly = AssemblyLoadContext.Default.LoadFromStream(stream);
#endif
               return assembly.GetExportedTypes();
}
```

After checking the [AssemblyLoadContext](https://github.com/dotnet/coreclr/blob/2c61b985c0513d9a70eee7221fa15910e7003d16/src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs), when calling LoadFromStream, first call VerifyIsAlive to check, and when the AssemblyLoadContext is destructed, it will set _state to InternalState.Unloading, which causes the AssemblyLoadContext.Default.LoadFromStream to throw an AssemblyLoadContext_Verify_NotUnloading exception. Can you help me solve it? ?
     Greetings
     Fanly

### PR DESCRIPTION
**Pull request by janvorli**
The OnProcessExit was iterating over the s_contextsToUnload and calling
Unload on each of the contexts. However the s_contextsToUnload is a
misnomer, it contains all the contexts, both collectible and
non-collectible ones. So we were calling Unload on non-collectible
contexts too there. That resulted in InvalidOperationException being
thrown from the OnProcessExit.

This change fixes that by calling unload only on ALCs whose
IsCollectible returns true. It also renames the s_contextsToUnload to
s_allContexts.

Close #22902

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs [hunk 1]
```diff
@@ -385,18 +385,18 @@ public void Unload()
...
         {
-            lock (s_contextsToUnload)
+            lock (s_allContexts)
             {
                 s_isProcessExiting = true;
-                foreach (var alcAlive in s_contextsToUnload)
+                foreach (var alcAlive in s_allContexts)
                 {
-                    if (alcAlive.Value.TryGetTarget(out AssemblyLoadContext alc))
+                    if (alcAlive.Value.TryGetTarget(out AssemblyLoadContext alc) && alc.IsCollectible)
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: Is this going to cause `Unload` event to not be delivered to regular contexts during shutdown? (diff from current behavior)

Comment (janvorli) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: Hmm, you are right, I think we were delivering those events for regular contexts in the past (even before the unloadability work started). That would be a breaking change. It seems we should call the Unload event for all contexts then (but not initiate the actual unloading for the non-collectible contexts)

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: We even have a test to verify this: 
System.Tests.UnloadingAndProcessExitTests.UnloadingEventMustHappenBeforeProcessExitEvent

Comment (janvorli) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: Actually, I've just found that the `InitiateUnload` when called from the `OnProcessExit` just sends the Unload event and sets `_state = InternalState.Unloading`. It doesn't initiate the unloading itself. 
So the problematic thing was actually setting the _state even though we are not unloading the context.

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs [hunk 2]
```diff
@@ -385,18 +383,17 @@ public void Unload()
...
                 {
                     if (alcAlive.Value.TryGetTarget(out AssemblyLoadContext alc))
                     {
-                        // Should we use a try/catch?
-                        alc.InitiateUnload();
+                        alc.RaiseUnloadEvent();
                     }
                 }
-                s_contextsToUnload.Clear();
+                s_allContexts.Clear();
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: Do we really need to call `Clear` here?

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs [hunk 3]
```diff
@@ -106,36 +106,34 @@ private protected AssemblyLoadContext(bool representsTPALoadContext, bool isColl
             }
         }
 
+        private void RaiseUnloadEvent()
+        {
+            // Ensure that we raise the Unload event only once
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: Note that there are subtle races where the unload event may not be delivered. Consider:

1. Unloadable context gets here, swaps in null, and gets rescheduled
2. Actual process Exit happens, delivers the event to all contexts (missing the one above) and process exits

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: (Not sure what to do about it.)

Comment (janvorli) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: That's intended behavior - or do you think we should send the unload event twice in such case?

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs (hunk 2 above)
Comment (janvorli) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: I think we don't have to. The process is going down and so we don't really care. In fact, I was thinking about this stuff over the weekend and I think that we should also remove the `s_isProcessExiting` variable. It just prevents creation of new AssemblyLoadContext after the OnProcessExit is called. In the light of your current PR for removal of special handling of the runtime shutdown, we should not care.

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs (hunk 3 above)
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: I am not sure whether sending the event twice would be an improvement. I think we can leave it the way it is now.

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs (hunk 2 above)
Comment (janvorli) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: I have added a commit removing the s_isProcessExiting and the Clear function call. @jkotas can you please take a look again?

**Pull request merged**